Failed tests (11265): TestStoreZoneUpdateAndRangeSplit

The following test appears to have failed:

[#11265](https://circleci.com/gh/cockroachdb/cockroach/11265):

```
1      storage/queue.go:333
1      kv/txn_coord_sender.go:754
I0113 18:44:10.649528 956 stopper.go:264  draining; tasks left:
1      storage/queue.go:333
I0113 18:44:10.650399 956 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- FAIL: TestStoreZoneUpdateAndRangeSplit (0.03s)
	client_split_test.go:546: range max bytes mismatch, got: 67108864, expected: 65536
=== RUN   TestStoreRangeSplitWithMaxBytesUpdate
I0113 18:44:10.651809 956 gossip/gossip.go:218  setting node descriptor node_id:1 address:<network_field:"localhost" address_field:"1" > attrs:<> 
W0113 18:44:10.653106 956 gossip/gossip.go:445  bad address from gossip node {1 1 }: gossip/resolver/resolver.go:96: unknown address network "localhost" for 1
I0113 18:44:10.655354 956 raft/raft.go:446  [group 1] 1 became follower at term 5
I0113 18:44:10.655610 956 raft/raft.go:234  [group 1] newRaft 1 [peers: [1], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0113 18:44:10.655766 956 raft/raft.go:526  [group 1] 1 is starting a new election at term 5
I0113 18:44:10.655882 956 raft/raft.go:459  [group 1] 1 became candidate at term 6
I0113 18:44:10.656011 956 raft/raft.go:508  [group 1] 1 received vote from 1 at term 6
I0113 18:44:10.656143 956 raft/raft.go:487  [group 1] 1 became leader at term 6
--
I0113 18:44:16.125046 956 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestLogSplits (1.02s)
	<autogenerated>:28: prior leak detected, leaktest disabled
=== RUN   Example_rebalancing
--- PASS: Example_rebalancing (0.90s)
FAIL
FAIL	github.com/cockroachdb/cockroach/storage	28.539s
=== RUN   TestBatchBasics
I0113 18:44:21.076334 1027 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.01s)
=== RUN   TestBatchGet
I0113 18:44:21.080807 1027 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.01s)
=== RUN   TestBatchMerge
I0113 18:44:21.091995 1027 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.00s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

--- a/cli/cli_test.go
+++ b/cli/cli_test.go
@@ -399,6 +399,8 @@ range_max_bytes: 67108864
 	// - attrs: [us-west-1b, ssd]
 	// range_min_bytes: 8388608
 	// range_max_bytes: 67108864
+	// gc:
+	//   ttlseconds: 0
 	//
 	// zone get 100
 	// replicas:
@@ -407,6 +409,8 @@ range_max_bytes: 67108864
 	// - attrs: [us-west-1b, ssd]
 	// range_min_bytes: 8388608
 	// range_max_bytes: 67108864
+	// gc:
+	//   ttlseconds: 0
 	//
 	// zone rm 100
 	// OK

--- a/config/config.go
+++ b/config/config.go
@@ -46,7 +46,7 @@ var (
 		},
 		RangeMinBytes: 1048576,
 		RangeMaxBytes: 67108864,
-		GC: &GCPolicy{
+		GC: GCPolicy{
 			TTLSeconds: 24 * 60 * 60, // 1 day
 		},
 	}

--- a/config/config.pb.go
+++ b/config/config.pb.go
@@ -53,11 +53,11 @@ type ZoneConfig struct {
 	// for each replica in the zone. The order in which the attributes are stored
 	// in ReplicaAttrs is arbitrary and may change.
 	ReplicaAttrs  []cockroach_roachpb.Attributes `protobuf:"bytes,1,rep,name=replica_attrs" json:"replica_attrs" yaml:"replicas,omitempty"`
-	RangeMinBytes int64                          `protobuf:"varint,2,opt,name=range_min_bytes" json:"range_min_bytes" yaml:"range_min_bytes,omitempty"`
-	RangeMaxBytes int64                          `protobuf:"varint,3,opt,name=range_max_bytes" json:"range_max_bytes" yaml:"range_max_bytes,omitempty"`
+	RangeMinBytes int64                          `protobuf:"varint,2,opt,name=range_min_bytes" json:"range_min_bytes" yaml:"range_min_bytes"`
+	RangeMaxBytes int64                          `protobuf:"varint,3,opt,name=range_max_bytes" json:"range_max_bytes" yaml:"range_max_bytes"`
 	// If GC policy is not set, uses the next highest, non-null policy
 	// in the zone config hierarchy, up to the default policy if necessary.
-	GC *GCPolicy `protobuf:"bytes,4,opt,name=gc" json:"gc,omitempty" yaml:"gc,omitempty"`
+	GC GCPolicy `protobuf:"bytes,4,opt,name=gc" json:"gc"`
 }
 
 func (m *ZoneConfig) Reset()         { *m = ZoneConfig{} }
@@ -131,16 +131,14 @@ func (m *ZoneConfig) MarshalTo(data []byte) (int, error) {
 	data[i] = 0x18
 	i++
 	i = encodeVarintConfig(data, i, uint64(m.RangeMaxBytes))
-	if m.GC != nil {
-		data[i] = 0x22
-		i++
-		i = encodeVarintConfig(data, i, uint64(m.GC.Size()))
-		n1, err := m.GC.MarshalTo(data[i:])
-		if err != nil {
-			return 0, err
-		}
-		i += n1
+	data[i] = 0x22
+	i++
+	i = encodeVarintConfig(data, i, uint64(m.GC.Size()))
+	n1, err := m.GC.MarshalTo(data[i:])
+	if err != nil {
+		return 0, err
 	}
+	i += n1
 	return i, nil
 }
 
@@ -219,10 +217,8 @@ func (m *ZoneConfig) Size() (n int) {
 	}
 	n += 1 + sovConfig(uint64(m.RangeMinBytes))
 	n += 1 + sovConfig(uint64(m.RangeMaxBytes))
-	if m.GC != nil {
-		l = m.GC.Size()
-		n += 1 + l + sovConfig(uint64(l))
-	}
+	l = m.GC.Size()
+	n += 1 + l + sovConfig(uint64(l))
 	return n
 }
 
@@ -443,9 +439,6 @@ func (m *ZoneConfig) Unmarshal(data []byte) error {
 			postIndex := iNdEx + msglen
 			if postIndex > l {
 				return io.ErrUnexpectedEOF
-			}
-			if m.GC == nil {
-				m.GC = &GCPolicy{}
 			}
 			if err := m.GC.Unmarshal(data[iNdEx:postIndex]); err != nil {
 				return err

--- a/config/config.proto
+++ b/config/config.proto
@@ -41,11 +41,11 @@ message ZoneConfig {
   // for each replica in the zone. The order in which the attributes are stored
   // in ReplicaAttrs is arbitrary and may change.
   repeated roachpb.Attributes replica_attrs = 1 [(gogoproto.nullable) = false, (gogoproto.moretags) = "yaml:\"replicas,omitempty\""];
-  optional int64 range_min_bytes = 2 [(gogoproto.nullable) = false, (gogoproto.moretags) = "yaml:\"range_min_bytes,omitempty\""];
-  optional int64 range_max_bytes = 3 [(gogoproto.nullable) = false, (gogoproto.moretags) = "yaml:\"range_max_bytes,omitempty\""];
+  optional int64 range_min_bytes = 2 [(gogoproto.nullable) = false, (gogoproto.moretags) = "yaml:\"range_min_bytes\""];
+  optional int64 range_max_bytes = 3 [(gogoproto.nullable) = false, (gogoproto.moretags) = "yaml:\"range_max_bytes\""];
   // If GC policy is not set, uses the next highest, non-null policy
   // in the zone config hierarchy, up to the default policy if necessary.
-  optional GCPolicy gc = 4 [(gogoproto.customname) = "GC", (gogoproto.moretags) = "yaml:\"gc,omitempty\""];
+  optional GCPolicy gc = 4 [(gogoproto.nullable) = false, (gogoproto.customname) = "GC"];
 }
 
 message SystemConfig {

--- a/roachpb/data.go
+++ b/roachpb/data.go
@@ -818,12 +818,17 @@ func AsIntents(spans []Span, txn *Transaction) []Intent {
 // TODO(tschottdorf): chance for code rot here. Adding a test like we did for
 // transactions (TestTransactionUpdate) can't hurt.
 func (s Span) Equal(o Span) bool {
-	return bytes.Equal(s.Key, o.Key) && bytes.Equal(s.EndKey, o.EndKey)
+	return s.Key.Equal(o.Key) && s.EndKey.Equal(o.EndKey)
 }
 
 // RSpan is a key range with an inclusive start RKey and an exclusive end RKey.
 type RSpan struct {
 	Key, EndKey RKey
+}
+
+// Equal compares for equality.
+func (rs RSpan) Equal(o RSpan) bool {
+	return rs.Key.Equal(o.Key) && rs.EndKey.Equal(o.EndKey)
 }
 
 // ContainsKey returns whether this span contains the specified key.

--- a/roachpb/metadata.go
+++ b/roachpb/metadata.go
@@ -104,23 +104,26 @@ func (a Attributes) SortedString() string {
 	return strings.Join(attrs, ",")
 }
 
+// RSpan returns the RangeDescriptor's resolved span.
+func (r RangeDescriptor) RSpan() RSpan {
+	return RSpan{Key: r.StartKey, EndKey: r.EndKey}
+}
+
 // ContainsKey returns whether this RangeDescriptor contains the specified key.
-func (r *RangeDescriptor) ContainsKey(key RKey) bool {
-	rs := RSpan{Key: r.StartKey, EndKey: r.EndKey}
-	return rs.ContainsKey(key)
+func (r RangeDescriptor) ContainsKey(key RKey) bool {
+	return r.RSpan().ContainsKey(key)
 }
 
 // ContainsKeyRange returns whether this RangeDescriptor contains the specified
 // key range from start (inclusive) to end (exclusive).
 // If end is empty, returns ContainsKey(start).
-func (r *RangeDescriptor) ContainsKeyRange(start, end RKey) bool {
-	rs := RSpan{Key: r.StartKey, EndKey: r.EndKey}
-	return rs.ContainsKeyRange(start, end)
+func (r RangeDescriptor) ContainsKeyRange(start, end RKey) bool {
+	return r.RSpan().ContainsKeyRange(start, end)
 }
 
 // FindReplica returns the replica which matches the specified store
 // ID. If no replica matches, (-1, nil) is returned.
-func (r *RangeDescriptor) FindReplica(storeID StoreID) (int, *ReplicaDescriptor) {
+func (r RangeDescriptor) FindReplica(storeID StoreID) (int, *ReplicaDescriptor) {
 	for i := range r.Replicas {
 		if r.Replicas[i].StoreID == storeID {
 			return i, &r.Replicas[i]
@@ -130,7 +133,7 @@ func (r *RangeDescriptor) FindReplica(storeID StoreID) (int, *ReplicaDescriptor)
 }
 
 // Validate performs some basic validation of the contents of a range descriptor.
-func (r *RangeDescriptor) Validate() error {
+func (r RangeDescriptor) Validate() error {
 	if r.NextReplicaID == 0 {
 		return util.Errorf("NextReplicaID must be non-zero")
 	}

--- a/storage/client_split_test.go
+++ b/storage/client_split_test.go
@@ -530,44 +530,40 @@ func TestStoreZoneUpdateAndRangeSplit(t *testing.T) {
 		t.Fatal(err)
 	}
 
-	// Wait for the range to be split along table boundaries.
-	originalRange := store.LookupReplica(roachpb.RKey(keys.UserTableDataMin), nil)
-	var rng *storage.Replica
-	util.SucceedsWithin(t, splitTimeout, func() error {
-		rng = store.LookupReplica(keys.MakeTablePrefix(descID), nil)
-		if rng.RangeID == originalRange.RangeID {
-			return util.Errorf("expected new range created by split")
+	tableBoundary := keys.MakeTablePrefix(descID)
+
+	{
+		var rng *storage.Replica
+
+		// Wait for the range to be split along table boundaries.
+		expectedRSpan := roachpb.RSpan{Key: roachpb.RKey(tableBoundary), EndKey: roachpb.RKeyMax}
+		util.SucceedsWithin(t, splitTimeout, func() error {
+			rng = store.LookupReplica(tableBoundary, nil)
+			if actualRSpan := rng.Desc().RSpan(); !actualRSpan.Equal(expectedRSpan) {
+				return util.Errorf("expected range %s to span %s", rng, expectedRSpan)
+			}
+			return nil
+		})
+
+		// Check range's max bytes settings.
+		if actualMaxBytes := rng.GetMaxBytes(); actualMaxBytes != maxBytes {
+			t.Fatalf("range %s max bytes mismatch, got: %d, expected: %d", rng, actualMaxBytes, maxBytes)
 		}
-		return nil
-	})
 
-	// Check range's max bytes settings.
-	if actualMaxBytes := rng.GetMaxBytes(); actualMaxBytes != maxBytes {
-		t.Fatalf("range %s max bytes mismatch, got: %d, expected: %d", rng, actualMaxBytes, maxBytes)
+		// Look in the range after prefix we're writing to.
+		fillRange(store, rng.RangeID, tableBoundary, maxBytes, t)
 	}
-
-	// Make sure the second range goes to the end.
-	if desc := rng.Desc(); !desc.EndKey.Equal(roachpb.RKeyMax) {
-		t.Fatalf("second range has split: %+v", desc)
-	}
-
-	// Look in the range after prefix we're writing to.
-	fillRange(store, rng.RangeID, keys.MakeTablePrefix(descID), maxBytes, t)
 
 	// Verify that the range is in fact split.
-	var newRng *storage.Replica
 	util.SucceedsWithin(t, splitTimeout, func() error {
-		newRng = store.LookupReplica(keys.MakeTablePrefix(descID+1), nil)
-		if newRng.RangeID == rng.RangeID {
-			return util.Errorf("range has not yet split")
+		rng := store.LookupReplica(keys.MakeTablePrefix(descID+1), nil)
+		rngDesc := rng.Desc()
+		rngStart, rngEnd := rngDesc.StartKey, rngDesc.EndKey
+		if rngStart.Equal(tableBoundary) || !rngEnd.Equal(roachpb.RKeyMax) {
+			return util.Errorf("range %s has not yet split", rng)
 		}
 		return nil
 	})
-
-	// Make sure the new range goes to the end.
-	if desc := newRng.Desc(); !desc.EndKey.Equal(roachpb.RKeyMax) {
-		t.Fatalf("second range has split: %+v", desc)
-	}
 }
 
 // TestStoreRangeSplitWithMaxBytesUpdate tests a scenario where a new

--- a/storage/client_split_test.go
+++ b/storage/client_split_test.go
@@ -542,20 +542,19 @@ func TestStoreZoneUpdateAndRangeSplit(t *testing.T) {
 	})
 
 	// Check range's max bytes settings.
-	if rng.GetMaxBytes() != maxBytes {
-		t.Fatalf("range max bytes mismatch, got: %d, expected: %d", rng.GetMaxBytes(), maxBytes)
+	if actualMaxBytes := rng.GetMaxBytes(); actualMaxBytes != maxBytes {
+		t.Fatalf("range %s max bytes mismatch, got: %d, expected: %d", rng, actualMaxBytes, maxBytes)
 	}
 
 	// Make sure the second range goes to the end.
-	if !roachpb.RKeyMax.Equal(rng.Desc().EndKey) {
-		t.Fatalf("second range has split: %+v", rng.Desc())
+	if desc := rng.Desc(); !desc.EndKey.Equal(roachpb.RKeyMax) {
+		t.Fatalf("second range has split: %+v", desc)
 	}
 
 	// Look in the range after prefix we're writing to.
 	fillRange(store, rng.RangeID, keys.MakeTablePrefix(descID), maxBytes, t)
 
-	// Verify that the range is in fact split (give it a few seconds for very
-	// slow test machines).
+	// Verify that the range is in fact split.
 	var newRng *storage.Replica
 	util.SucceedsWithin(t, splitTimeout, func() error {
 		newRng = store.LookupReplica(keys.MakeTablePrefix(descID+1), nil)
@@ -566,8 +565,8 @@ func TestStoreZoneUpdateAndRangeSplit(t *testing.T) {
 	})
 
 	// Make sure the new range goes to the end.
-	if !roachpb.RKeyMax.Equal(newRng.Desc().EndKey) {
-		t.Fatalf("second range has split: %+v", rng.Desc())
+	if desc := newRng.Desc(); !desc.EndKey.Equal(roachpb.RKeyMax) {
+		t.Fatalf("second range has split: %+v", desc)
 	}
 }
 

--- a/storage/gc_queue.go
+++ b/storage/gc_queue.go
@@ -27,6 +27,7 @@ import (
 	"github.com/cockroachdb/cockroach/keys"
 	"github.com/cockroachdb/cockroach/roachpb"
 	"github.com/cockroachdb/cockroach/storage/engine"
+	"github.com/cockroachdb/cockroach/util"
 	"github.com/cockroachdb/cockroach/util/log"
 	"github.com/gogo/protobuf/proto"
 )
@@ -97,13 +98,12 @@ func (*gcQueue) shouldQueue(now roachpb.Timestamp, repl *Replica,
 	desc := repl.Desc()
 	zone, err := sysCfg.GetZoneConfigForKey(desc.StartKey)
 	if err != nil {
-		log.Errorf("could not find GC policy for range %s: %s", repl, err)
+		log.Errorf("could not find zone config for range %s: %s", repl, err)
 		return
 	}
-	policy := zone.GC
 
 	// GC score is the total GC'able bytes age normalized by 1 MB * the replica's TTL in seconds.
-	gcScore := float64(repl.stats.GetGCBytesAge(now.WallTime)) / float64(policy.TTLSeconds) / float64(gcByteCountNormalization)
+	gcScore := float64(repl.stats.GetGCBytesAge(now.WallTime)) / float64(zone.GC.TTLSeconds) / float64(gcByteCountNormalization)
 
 	// Intent score. This computes the average age of outstanding intents
 	// and normalizes.
@@ -155,11 +155,10 @@ func (gcq *gcQueue) process(now roachpb.Timestamp, repl *Replica,
 	// Lookup the GC policy for the zone containing this key range.
 	zone, err := sysCfg.GetZoneConfigForKey(desc.StartKey)
 	if err != nil {
-		return fmt.Errorf("could not find GC policy for range %s: %s", repl, err)
+		return util.Errorf("could not find zone config for range %s: %s", repl, err)
 	}
-	policy := zone.GC
 
-	gc := engine.NewGarbageCollector(now, *policy)
+	gc := engine.NewGarbageCollector(now, zone.GC)
 
 	// Compute intent expiration (intent age at which we attempt to resolve).
 	intentExp := now

--- a/util/http_test.go
+++ b/util/http_test.go
@@ -47,6 +47,8 @@ var yamlConfig = []byte(`replicas:
 - attrs: [b, hdd]
 range_min_bytes: 1048576
 range_max_bytes: 67108864
+gc:
+  ttlseconds: 0
 `)
 
 var jsonConfig = []byte(`{
@@ -77,7 +79,10 @@ var jsonConfig = []byte(`{
     }
   ],
   "range_min_bytes": 1048576,
-  "range_max_bytes": 67108864
+  "range_max_bytes": 67108864,
+  "gc": {
+    "ttl_seconds": 0
+  }
 }`)
 
 var protobufConfig []byte
@@ -138,7 +143,7 @@ func TestUnmarshalRequest(t *testing.T) {
 				t.Errorf("%d: unexpected success", i)
 			}
 			continue
-		} else if !test.expError && err != nil {
+		} else if err != nil {
 			t.Errorf("%d: unexpected failure: %s", i, err)
 			continue
 		}
@@ -186,7 +191,7 @@ func TestMarshalResponse(t *testing.T) {
 			t.Fatalf("%d: %s", i, err)
 		}
 		if !bytes.Equal(body, test.expBody) {
-			t.Errorf("%d: expected %q; got %q", i, test.expBody, body)
+			t.Errorf("%d: expected:\n%q\ngot\n%q", i, test.expBody, body)
 		}
 		if cType != test.expCType {
 			t.Errorf("%d: expected %s content type; got %s", i, test.expCType, cType)

